library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,568 @@
+    end
+
+    # generates description for CWM with widget. Description is auto filled from defined methods.
+    #
+    # methods used to generate description:
+    #
+    # - `#help` [String] to get translated help text for widget
+    # - `#label` [String] to get translated label text for widget
+    # - `#opt` [Array<Symbol>] to get options passed to widget like `[:hstretch, :vstretch]`
+    # - `#validate` [Boolean ()] validate widget value. Returns false if validation failed.
+    # - `#init` [nil ()] initialize widget like e.g. set initial value
+    # - `#handle` [Symbol,nil (String?)] handle widget changed value or press.
+    #   return value is usually nil, returning symbol can be used to send different event.
+    #   It support varient with parameter specified and without. If parameter is specified,
+    #   gets event Hash including "ID" key with value that specify widget ID that cause event.
+    #   It is mainly useful if handle_all_events is set to true to distinguish which event caused it.
+    # - `#store` [nil ()] store widget value after user confirm dialog
+    # - `#cleanup` [nil ()] cleanup after widget is destroyed
+    # @raise [RuntimeError] if required method is not implemented or widget id not set.
+    def description
+      if !respond_to?(:widget_type)
+        raise "Widget '#{self.class}' does set its widget type"
+      end
+
+      res = {}
+
+      if respond_to?(:help)
+        res["help"] = help
+      else
+        res["no_help"] = ""
+      end
+      res["label"] = label if respond_to?(:label)
+      res["opt"] = opt if respond_to?(:opt)
+      if respond_to?(:validate)
+        res["validate_function"] = validate_method
+        res["validate_type"] = :function
+      end
+      res["handle_events"] = [widget_id] unless handle_all_events
+      res["init"] = init_method if respond_to?(:init)
+      res["handle"] = handle_method if respond_to?(:handle)
+      res["store"] = store_method if respond_to?(:store)
+      res["cleanup"] = cleanup_method if respond_to?(:cleanup)
+      res["widget"] = widget_type
+
+      res
+    end
+
+    # gets if widget is open for modification
+    def enabled?
+      Yast::UI.QueryWidget(Id(widget_id), :Enabled)
+    end
+
+    # Opens widget for modification
+    def enable
+      Yast::UI.ChangeWidget(Id(widget_id), :Enabled, true)
+    end
+
+    # Closes widget for modification
+    def disable
+      Yast::UI.ChangeWidget(Id(widget_id), :Enabled, false)
+    end
+
+  protected
+
+    # helper to check if event is invoked by this widget
+    def my_event?(event)
+      widget_id == event["ID"]
+    end
+
+    # shortcut from Yast namespace to avoid including whole namespace
+    # TODO: kill converts in CWM module, to avoid this workaround for funrefs
+    def fun_ref(*args)
+      Yast::FunRef.new(*args)
+    end
+
+  private
+
+    # @note all methods here use wrappers to modify required parameters as CWM
+    # have not so nice callbacks API
+    def init_method
```

Comment:
Non-API nitpicking, but it irritates me :)
`def init_fun_ref`
One day it may return a Method, but now it takes a Method and turns it into a FunRef, so it should have a name that does not lie.